### DISCUSSION TITLE
Different streams get returned using the same inputs in different environments

### DISCUSSION THREAD

**Discussion by campones** [streamlink/streamlink#4758] [Q&A] [answered]
I m facing some issue with Tf1 plugin, at least one of the channel ( tf1 filme et series), I don't have sound when selecting the regular streams available ( ex 576p)  

So when I m on a terminal  and I want to see the available streams I see this

/usr/local/bin/streamlink "https://www.tf1.fr/hd1/direct"
[cli][info] streamlink is running as root! Be careful!
[cli][info] Found matching plugin tf1 for URL https://www.tf1.fr/hd1/direct
Available streams: 720p_hls, 576p_hls, 360p_hls, 234p_hls, 234p (worst), 360p, 576p_alt, 576p, 720p (best)

When I m launching the same stream via a cronjob, in the cronjob logs I see this instead

[cli][info] streamlink is running as root! Be careful!
[cli][info] Found matching plugin tf1 for URL https://www.tf1.fr/hd1/direct
error: The specified stream(s) '576p_hls' could not be found.
       Available streams: 720p_dash, 576p_dash, 576p_dash_alt, 360p_dash, 234p_dash, 234p (worst), 360p, 576p_alt, 576p, 720p (best)

This seems pretty weird no? I mean, it's the same streamlink version (last one) same server , same everything.

**Comment by bastimeyer** [comment, marked as answer]
> This seems pretty weird no?

No.

The TF1 plugin returns both `hls-multi` (MuxedHLSStream) as well as `dash` (DASHStream) streams. If there are streams with the same name, then there can be name collisions once all `_alt` stream slots are consumed. This is a limitation of the stream collection design of Streamlink. I have talked about this a couple of times, but changing it would be a really huge undertaking that's probably not worth the effort, as it affects not just streamlink and streamlink_cli, but literally every single plugin and how they return/yield streams.

Stream types can be limited by the `--stream-types` CLI argument if you only want certain types of streams or if you want to change the stream type priority.
https://streamlink.github.io/cli.html#cmdoption-stream-types

The plugin shouldn't return both stream types and should prefer HLS streams, if available. As you can see in the plugin's git history, it hasn't been updated in a long time.

The reason why you're seeing different results is because one of your environment can't resolve the `ffmpeg` binary via the `PATH` env var, so `hls-multi` streams fall back to `hls` without audio.

```
$ streamlink --json 'https://www.tf1.fr/hd1/direct' \
  | jq -r '.streams | to_entries | .[] | "\(.key): \(.value.type)"'
720p_hls: hls-multi
576p_hls: hls-multi
360p_hls: hls-multi
234p_hls: hls-multi
234p: dash
360p: dash
576p_alt: dash
576p: dash
720p: dash
worst: dash
best: dash
```

```
$ streamlink --stream-types 'hls-multi,*' --json 'https://www.tf1.fr/hd1/direct' \
  | jq -r '.streams | to_entries | .[] | "\(.key): \(.value.type)"'
720p_dash: dash
576p_dash: dash
576p_dash_alt: dash
360p_dash: dash
234p_dash: dash
234p: hls-multi
360p: hls-multi
576p_alt: hls-multi
576p: hls-multi
720p: hls-multi
worst: hls-multi
best: hls-multi

$ streamlink --stream-types 'dash,*' --json 'https://www.tf1.fr/hd1/direct' \
  | jq -r '.streams | to_entries | .[] | "\(.key): \(.value.type)"'
720p_hls: hls-multi
576p_hls: hls-multi
360p_hls: hls-multi
234p_hls: hls-multi
234p: dash
360p: dash
576p_alt: dash
576p: dash
720p: dash
worst: dash
best: dash
```

```
$ streamlink --ffmpeg-ffmpeg=/dev/null --json 'https://www.tf1.fr/hd1/direct' \
  | jq -r '.streams | to_entries | .[] | "\(.key): \(.value.type)"'
720p_dash: dash
576p_dash: dash
576p_dash_alt: dash
360p_dash: dash
234p_dash: dash
234p: hls
360p: hls
576p_alt: hls
576p: hls
720p: hls
worst: hls
best: hls
```

**Comment by campones** [comment]
thank you for such a detailed answer.
Would you happen to know how I can add the ffmpeg binary PATH to the ENV
I have previously noticed that streamlink (and therefore ffmpeg) is able to remux the hls multi when I launch this in command line but not while using the cronjob., despite the cronjob is run as root.

**Reply by bastimeyer** [reply]
Out of scope. Do your own research.

- https://www.google.com/search?q=linux+update+path+environment+variable
- https://wiki.archlinux.org/title/Environment_variables

**Reply by campones** [reply]
yes, I m researching about this now, thanks

**Reply by bastimeyer** [reply]
Also
https://streamlink.github.io/cli.html#cmdoption-ffmpeg-ffmpeg

**Reply by campones** [reply]
yes, this last option is a great work around to make this instead to have to change the crontab environment path,  so I can say, thank you for fixing my personal issue!  It's working already
![2e1cbcc65b55fdb81ebfa7a96becac18](https://user-images.githubusercontent.com/7072524/185762820-a76f2622-2543-4b2c-ad28-856750ed6de1.png)

**Comment by campones** [comment]
Well as I suspected , all the tf1 group streaming have now moved to this hls-multi

**Comment by campones** [comment]
hello again

So mostly your answer was that because of different environment (more specificaly ffmpeg path was missing while using a cron) it would give a different output in term of streams' availability.  Which is very true .
There is still something I don't understand, it's really messy in my head I reckon.. 

So I have 2 ubuntu 20.04 servers I have downloaded a static ffmpeg 4.4.3 binary, and I m using this now in the --ffmpeg-ffmpeg option.
I can't understand why while using the same command in the terminal,  on one server chosing 576p will give me a dash stream, and on the other an hls-multi stream. 

server 1 
```
/usr/local/bin/streamlink --ffmpeg-ffmpeg "/usr/local/bin/ffmpeg4" --ffmpeg-audio-transcode "aac" --ffmpeg-copyts --ffmpeg-fout "flv" "https://www.tf1.fr/tf1/direct" 576p
[cli][info] streamlink is running as root! Be careful!
[cli][info] Found matching plugin tf1 for URL https://www.tf1.fr/tf1/direct
[cli][info] Available streams: 234p (worst), 360p, 576p_alt, 576p, 720p (best)
[cli][info] Opening stream: 576p (hls-multi)
[cli][info] Starting player: /snap/bin/vlc
```

server 2 
```
/usr/local/bin/streamlink --ffmpeg-ffmpeg "/usr/local/bin/ffmpeg4" --ffmpeg-audio-transcode "aac" --ffmpeg-copyts --ffmpeg-fout "flv" "https://www.tf1.fr/tf1/direct" 576p
[cli][info] streamlink is running as root! Be careful!
[cli][info] Found matching plugin tf1 for URL https://www.tf1.fr/tf1/direct
[cli][info] Available streams: 720p_hls, 576p_hls, 360p_hls, 234p_hls, 234p (worst), 360p, 576p_alt, 576p, 720p (best)
[cli][info] Opening stream: 576p (dash)
```

edit:  on one server I still had streamlink 4.3 and 5.1 on the other one, after updating I get the same output.  .

**Comment by campones** [comment]
I have to say that since I upgraded my main server with streamlink 5.1 I do not suffer from those weird out of sync issues anymore while using hls-multi. That seems to prove that something was buggy in previous releases.
